crates/server/src/access.rs

```diff
@@ -0,0 +1,182 @@
+use std::net::IpAddr;
+
+use hickory_proto::error::{ProtoError, ProtoErrorKind};
+use ipnet::{IpNet, Ipv4Net, Ipv6Net};
+use prefix_trie::{Prefix, PrefixSet};
+
+/// Type to evaluate access from a source address for accessing the server.
+///
+/// Allowed networks will override denied networks, i.e. if a network is allowed, the deny rules will not be evaluated.
+/// Allowed networks are processed in the context of denied networks, that is, if there are no denied networks, then
+///   the allowed list will effectively deny access to anything that's not in the allowed list. On the other hand, if
+///   denied networks are specified, then allowed networks will only apply if the deny rule matched, but otherwise the
+///   address will be allowed.
+#[derive(Default)]
+pub(crate) struct Access {
+    deny_ipv4: Option<PrefixSet<Ipv4Net>>,
+    deny_ipv6: Option<PrefixSet<Ipv6Net>>,
+    allow_ipv4: Option<PrefixSet<Ipv4Net>>,
+    allow_ipv6: Option<PrefixSet<Ipv6Net>>,
+}
+
+impl Access {
+    /// Insert a new network that is denied access to the server
+    pub(crate) fn insert_deny(&mut self, network: IpNet) {
+        match network {
+            IpNet::V4(v4) => {
+                self.deny_ipv4
+                    .get_or_insert_with(PrefixSet::default)
+                    .insert(v4);
+            }
+            IpNet::V6(v6) => {
+                self.deny_ipv6
+                    .get_or_insert_with(PrefixSet::default)
+                    .insert(v6);
+            }
+        }
+    }
+
+    /// Insert a new network that is allowed access to the server
+    pub(crate) fn insert_allow(&mut self, network: IpNet) {
+        match network {
+            IpNet::V4(v4) => {
+                self.allow_ipv4
+                    .get_or_insert_with(PrefixSet::default)
+                    .insert(v4);
+            }
+            IpNet::V6(v6) => {
+                self.allow_ipv6
+                    .get_or_insert_with(PrefixSet::default)
+                    .insert(v6);
+            }
+        }
+    }
+
+    pub(crate) fn insert_deny_all(&mut self, networks: &[IpNet]) {
+        for net in networks {
+            self.insert_deny(*net);
+        }
+    }
+
+    pub(crate) fn insert_allow_all(&mut self, networks: &[IpNet]) {
+        for net in networks {
+            self.insert_allow(*net);
+        }
+    }
+
+    /// Evaluate the IP address against the allowed networks
+    ///
+    /// # Return
+    ///
+    /// Ok if access is granted, Err otherwise
+    pub(crate) fn allow(&self, ip: IpAddr) -> Result<(), ProtoError> {
```

Comment:
Nit: I would probably just return `bool` here, which seems simpler. The caller can turn that into `ProtoError` (maybe there could be a method like `ProtoError::request_refused()` for that?).

Reply:
Yeah, I considered that. The reason I decided on Result is to ensure it's evaluated. The other option is `Result<(), ()>`, I'd be more comfortable with that? Preference?

Reply:
Alternatively maybe use `bool` but with a `#[must_use]` attribute on the function?